### PR TITLE
Anonymous scan enhancements & cleanup

### PR DESCRIPTION
**Pull request by universalmind303**
general cleanup on the anonymousscan trait. 

i removed the `skip_rows` and `n_rows` from the scan_args because they can be pushed down through the lazyframe, so they are unnecessary. 

also added a few methods to the trait to allow the implementor to specify what should be pushed down. (predicate, pushdown, slice) being the supported pushdowns. 

added a few comments to the logical & physical plans explaining what they do, as I had some trouble initially understanding them. 

also small cleanup in the row infer so anonymousscan implementors can leverage `infer_schema` and `coerce_data_type` easier.

### CODE REVIEW

**Comment by ritchie46** [comment]
>  removed the skip_rows and n_rows

I think those should remain. The optimizer still has to store the pushdown state in the node.

**Comment by universalmind303** [comment]
> > removed the skip_rows and n_rows
> 
> I think those should remain. The optimizer still has to store the pushdown state in the node.
> 
> The implementation of the node can then use that state to apply it efficiently.

so the options are still part of the optimizer, but they are not passed directly in from the implementor anymore & use the lazy api (which i feel is a bit more idiomatic). If you think there is a use-case to have them in both, i can add it back. 

```rust
// the options used by the optimizer still has them both. 
pub struct AnonymousScanOptions {
    pub schema: SchemaRef,
    pub output_schema: Option<SchemaRef>,
    pub skip_rows: Option<usize>,
    pub n_rows: Option<usize>,
    pub with_columns: Option<Arc<Vec<String>>>,
    pub(crate) fmt_str: &'static str,
}
```

initially I had them added to the implementing interface & the lazy api, so you could set `n_rows` twice without realizing it. 
```rust

// before
let args = ScanArgsAnonymous { n_rows: Some(10), ..ScanArgsAnonymous::default()};
let lf = LazyFrame::anonymous_scan(get_df_somehow, args)
  .limit(20); // will only return 10 because of the first `n_rows`

// after

let lf = LazyFrame::anonymous_scan(get_df_somehow, ScanArgsAnonymous::default())
  .limit(10);

```

**Comment by ritchie46** [comment]
> so the options are still part of the optimizer, but they are not passed directly in from the implementor anymore & use the lazy api (which i feel is a bit more idiomatic). If you think there is a use-case to have them in both, i can add it back.

I agree that it is more clean to let the optimizer handle the pushdown. But still it would be valuable to have the option. For instance we want `scan_csv` to have mostly the same arguments as `read_csv`.

**Comment by universalmind303** [comment]
@ritchie46, I added both of those back in. Let me know if you have any other recommendations!

File: polars/polars-lazy/src/logical_plan/anonymous_scan.rs [hunk 1]
```diff
@@ -3,12 +3,35 @@ use polars_core::prelude::*;
...
+    /// Allows for an optional infer schema argument for data sources with dynamic schemas
     fn schema(&self, _infer_schema_length: Option<usize>) -> Result<Schema> {
         Err(PolarsError::ComputeError(
             "Must supply either a schema or a schema function".into(),
         ))
     }
+    /// specify if the scan provider should allow predicate pushdowns
+    ///
+    /// Defaults to `false`
+    fn allows_predicate_pushdown(&self) -> Result<bool> {
```
Comment (ritchie46) [inline polars/polars-lazy/src/logical_plan/anonymous_scan.rs:19]: Should this be fallible? I think we can just return `bool`.

**Comment by ritchie46** [comment]
I have a small nit and then I will merge it in. :+1:

**Pull request merged**